File: components/credentials/src/test/java/org/cloudfoundry/credhub/services/CredentialDataServiceTest.java
package org.cloudfoundry.credhub.services;

import java.util.UUID;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.junit4.SpringRunner;
import org.springframework.transaction.annotation.Transactional;

import org.cloudfoundry.credhub.CredhubTestApp;
import org.cloudfoundry.credhub.audit.CEFAuditRecord;
import org.cloudfoundry.credhub.entity.Credential;
import org.cloudfoundry.credhub.repositories.CredentialRepository;
import org.cloudfoundry.credhub.utils.DatabaseProfileResolver;
import org.junit.Test;
import org.junit.runner.RunWith;

import static org.hamcrest.CoreMatchers.instanceOf;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.core.IsEqual.equalTo;

@RunWith(SpringRunner.class)
@ActiveProfiles(value = "unit-test", resolver = DatabaseProfileResolver.class)
@SpringBootTest(classes = CredhubTestApp.class)
@Transactional
public class CredentialDataServiceTest {
  private static final String CREDENTIAL_NAME = "/test/credential";

  @Autowired
  private CredentialDataService subject;

  @Autowired
  private CredentialRepository credentialRepository;

  @Autowired
  private CEFAuditRecord auditRecord;

  @Test
  public void save_savesTheCredential() {
    final Credential credential = new Credential(CREDENTIAL_NAME);

    assertThat(credentialRepository.count(), equalTo(0L));

    credentialRepository.save(credential);

    assertThat(credentialRepository.count(), equalTo(1L));

    assertThat(
      credentialRepository.findOneByNameIgnoreCase(CREDENTIAL_NAME).getName(),
      equalTo(CREDENTIAL_NAME)
    );
  }

  @Test
  public void save_setsTheUuidOnTheCredential() {
    final Credential credential = credentialRepository.save(new Credential(CREDENTIAL_NAME));

    assertThat(credential.getUuid(), instanceOf(UUID.class));
  }

  @Test
  public void find_whenTheCredentialExists_returnsTheCredential() {
    final Credential credential = new Credential(CREDENTIAL_NAME);
    credentialRepository.save(credential);

    assertThat(subject.find(CREDENTIAL_NAME), equalTo(credential));
  }

  @Test
  public void find_isCaseInsensitive() {
    final Credential credential = new Credential(CREDENTIAL_NAME.toLowerCase());
    credentialRepository.save(credential);

    assertThat(subject.find(CREDENTIAL_NAME.toUpperCase()), equalTo(credential));
  }

  @Test
  public void find_whenTheCredentialDoesNotExist_returnsNull() {
    assertThat(subject.find(CREDENTIAL_NAME), equalTo(null));
  }

  @Test
  public void findByUUID_whenTheCredentialExists_returnsTheCredential() {
    final Credential credential = new Credential(CREDENTIAL_NAME);
    credentialRepository.save(credential);
    assertThat(subject.findByUUID(credential.getUuid()), equalTo(credential));
  }

  @Test
  public void delete_whenTheCredentialExists_deletesTheCredential_andReturnsTrue() {
    credentialRepository.save(new Credential(CREDENTIAL_NAME));

    assertThat(subject.delete(CREDENTIAL_NAME), equalTo(true));
    assertThat(credentialRepository.count(), equalTo(0L));
  }

  @Test
  public void delete_addsToAuditRecord() {
    credentialRepository.save(new Credential(CREDENTIAL_NAME));

    assertThat(subject.delete(CREDENTIAL_NAME), equalTo(true));
    assertThat(auditRecord.getResourceName(), is(CREDENTIAL_NAME));
  }

  @Test
  public void delete_whenTheCredentialDoesNotExist_returnsFalse() {
    assertThat(subject.delete(CREDENTIAL_NAME), equalTo(false));
  }

  @Test
  public void delete_isCaseInsensitive() {
    credentialRepository.save(new Credential(CREDENTIAL_NAME.toUpperCase()));

    assertThat(subject.delete(CREDENTIAL_NAME.toLowerCase()), equalTo(true));
    assertThat(credentialRepository.count(), equalTo(0L));
  }
}

File: components/credentials/src/test/kotlin/org/cloudfoundry/credhub/services/CredentialDataServiceTest.kt
package org.cloudfoundry.credhub.services

import org.cloudfoundry.credhub.CredhubTestApp
import org.cloudfoundry.credhub.audit.CEFAuditRecord
import org.cloudfoundry.credhub.entities.Credential
import org.cloudfoundry.credhub.repositories.CredentialRepository
import org.cloudfoundry.credhub.utils.DatabaseProfileResolver
import org.hamcrest.CoreMatchers
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.hamcrest.core.IsEqual
import org.junit.Test
import org.junit.runner.RunWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.context.junit4.SpringRunner
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@RunWith(SpringRunner::class)
@ActiveProfiles(value = ["unit-test"], resolver = DatabaseProfileResolver::class)
@SpringBootTest(classes = [CredhubTestApp::class])
@Transactional
class CredentialDataServiceTest {
    @Autowired
    private val subject: CredentialDataService? = null
    @Autowired
    private val credentialRepository: CredentialRepository? = null
    @Autowired
    private val auditRecord: CEFAuditRecord? = null

    @Test
    fun save_savesTheCredential() {
        val credential = Credential(CREDENTIAL_NAME)
        MatcherAssert.assertThat(credentialRepository!!.count(), IsEqual.equalTo(0L))
        credentialRepository.save(credential)
        MatcherAssert.assertThat(credentialRepository.count(), IsEqual.equalTo(1L))
        MatcherAssert.assertThat(
            credentialRepository.findOneByNameIgnoreCase(CREDENTIAL_NAME)!!.name,
            IsEqual.equalTo(CREDENTIAL_NAME)
        )
    }

    @Test
    fun save_setsTheUuidOnTheCredential() {
        val credential = credentialRepository!!.save(Credential(CREDENTIAL_NAME))
        MatcherAssert.assertThat(credential.uuid, CoreMatchers.instanceOf(UUID::class.java))
    }

    @Test
    fun find_whenTheCredentialExists_returnsTheCredential() {
        val credential = Credential(CREDENTIAL_NAME)
        credentialRepository!!.save(credential)
        MatcherAssert.assertThat(subject!!.find(CREDENTIAL_NAME), IsEqual.equalTo(credential))
    }

    @Test
    fun find_isCaseInsensitive() {
        val credential = Credential(CREDENTIAL_NAME.lowercase())
        credentialRepository!!.save(credential)
        MatcherAssert.assertThat(subject!!.find(CREDENTIAL_NAME.uppercase()), IsEqual.equalTo(credential))
    }

    @Test
    fun find_whenTheCredentialDoesNotExist_returnsNull() {
        MatcherAssert.assertThat(subject!!.find(CREDENTIAL_NAME), IsEqual.equalTo<Credential?>(null))
    }

    @Test
    fun findByUUID_whenTheCredentialExists_returnsTheCredential() {
        val credential = Credential(CREDENTIAL_NAME)
        credentialRepository!!.save(credential)
        MatcherAssert.assertThat(subject!!.findByUUID(credential.uuid), IsEqual.equalTo(credential))
    }

    @Test
    fun delete_whenTheCredentialExists_deletesTheCredential_andReturnsTrue() {
        credentialRepository!!.save(Credential(CREDENTIAL_NAME))
        MatcherAssert.assertThat(subject!!.delete(CREDENTIAL_NAME), IsEqual.equalTo(true))
        MatcherAssert.assertThat(credentialRepository.count(), IsEqual.equalTo(0L))
    }

    @Test
    fun delete_addsToAuditRecord() {
        credentialRepository!!.save(Credential(CREDENTIAL_NAME))
        MatcherAssert.assertThat(subject!!.delete(CREDENTIAL_NAME), IsEqual.equalTo(true))
        MatcherAssert.assertThat(auditRecord!!.resourceName, Matchers.`is`(CREDENTIAL_NAME))
    }

    @Test
    fun delete_whenTheCredentialDoesNotExist_returnsFalse() {
        MatcherAssert.assertThat(subject!!.delete(CREDENTIAL_NAME), IsEqual.equalTo(false))
    }

    @Test
    fun delete_isCaseInsensitive() {
        credentialRepository!!.save(Credential(CREDENTIAL_NAME.uppercase()))
        MatcherAssert.assertThat(subject!!.delete(CREDENTIAL_NAME.lowercase()), IsEqual.equalTo(true))
        MatcherAssert.assertThat(credentialRepository.count(), IsEqual.equalTo(0L))
    }

    companion object {
        private const val CREDENTIAL_NAME = "/test/credential"
    }
}
